src/main/com/fulcrologic/rad/form.cljc

```diff
@@ -138,17 +138,24 @@
         field-style (or (get field-styles qualified-key) field-style)]
     (if field-style
       (fn [env attr _] (render-field env attr))
-      (let [{::keys [ui layout-styles]} (get subforms qualified-key)
-            {target-styles ::layout-styles} (comp/component-options ui)
-            {::app/keys [runtime-atom]} (comp/any->app form-instance)
-            element      :ref-container
-            layout-style (or
-                           (get layout-styles element)
-                           (get target-styles element)
-                           :default)
-            render-fn    (some-> runtime-atom deref ::rad/controls ::element->style->layout
-                           (get-in [element layout-style]))]
-        render-fn))))
+      (let [{::keys [ui layout-styles subform-visible?]} (get subforms qualified-key)
+            subform-visible? (cond
```

Comment:
Ah, the RAD pattern is to use the utility function `?!` for this kind of thing. See line 1103 for an example (field-visible?).

Also, does this function need to be able to handle the to-many case? If so passing just the form-instance is insufficient.  I think the arguments should probably be the parent form (form-instance) and the props of the form that is being decided about, along with the join attribute (which could also hold the visible option if we want):

`(fn [parent-form join-attr child-props] boolean?)`

Well, I guess you could argue that what *you* need is the ability to hide the entire JOIN, but one might also want to hide a particular item in a list of subforms...both are "generically" valid desires.

Security seems like the ideal scenario where you might want this, in which case a to-many join might make individual decisions about each item.

Hiding a particular entire JOIN might just be considered a field-visible on the join key attribute.

Thoughts?

Reply:
Re `?!` - sure - done locally.

In the old days there were two types of relationships: lookup and master-detail. The lookups were the to-one case. A lookup is to something non-scalar, but which is still a kind of a 'field' - a field that would normally be on the same panel on the screen. So I'm going back to liking the idea of re-using the option `field-visible?`. Master-detail is a whole different thing - when does it make sense to hide the detail part dynamically? The detail part is usually on a separate panel - is usually structural and static. So no, I don't see a need to handle the to-many case, not in the near future anyway.

> "want to hide a particular item in a list of subforms..."

Why can't a form (whether subform or parent form) be responsible for hiding its own items/fields? To me it is the same argument - every form can be responsible for hiding its own fields, where a field is either a scalar or a lookup (a to-one).

> "Hiding a particular entire JOIN might just be considered a field-visible on the join key attribute."

So that's you thinking to keep the name `field-visible?`. If so the fact that the fn signature matches the `field-visible?` option on an attribute (which will be repeated for `fields-visible`) further enhances the case for this enhancement not going to the master-detail (to-many) case.

Reply:
Having said all that this is the `fo/subforms` where I'm using it:
```
{::bank-account/files {fo/ui idc/FileForm
                       fo/field-visible? (fn [this]
                                                (let [attrib-value (-> this comp/props ::bank-account/headings)
                                                      b? (attr/valid-value? bank-account.attribs/headings attrib-value)]
                                                  b?))
                       fo/title                 "Import Bank Statements"
                       fo/can-delete?           (fn [_ _] true)
                       fo/layout-styles         {:ref-container :file}
                       ::form/added-via-upload? true}}
```
I'm thinking of `::bank-account/files` as a lookup but it is a to-many ref. I can create more than one of these files and none of them are visible (the whole subform is invisible) when I make headings invalid. So it works the way I want it to.

Reply:
The problem with your current solution is that when the form has its own `ref-container-renderer` then `field-visible?` on the attribute is being ignored. If you keep my changes to fulcro-rad then `field-visible?` as a function is picked up. This is what I need - can't do without having the `ref-container-renderer` for `::bank-account/files`.